### PR TITLE
feat(bmm): add monorepo/multi-project support via skills

### PR DESCRIPTION
**Pull request by serrnovik**
## What
Implemented monorepo support for BMAD-METHOD, introducing a dynamic artifact pathing mechanism based on a local project context file (`.current_project`). This update allows workflows to automatically route their outputs to project-specific subdirectories within `_bmad-output/`.

This is tools based implementation of [PR based on legacy workflow](https://github.com/bmad-code-org/BMAD-METHOD/pull/1691) Unlike the previous approach, this implementation uses **skills** (`bmad-project-list`, `bmad-project-new`, `bmad-project-switch`) instead of a single `/set-project` workflow, aligning with the v6.1.0 tools-based architecture.
- Add `monorepo_context` config section to `module.yaml` for path resolution
- Add `bmad-project-list`, `bmad-project-new`, `bmad-project-switch` skills
- Config-level hook: LLM reads config and applies context before path vars
- Inline overrides via `#project:NAME` or `#p:NAME` in invocations
- Installer: ensure 0-context workflow dir is installed
- Add `docs/how-to/monorepo-setup.md`
- 100% backward compatible when no `.current_project` file

## Why
This change enables cleaner organization and branch switching in monorepo environments by centralizing methodology files while isolating project artifacts. It avoids methodology duplication and re-configuration overhead when working across multiple projects in a single shared repository.
## How
- **Config-level hook:** Added `monorepo_context` section to `module.yaml`. The LLM reads config and applies context resolution before using any path variables. Zero changes to existing workflow files.
- **Project management skills:** Created `bmad-project-list`, `bmad-project-new`, and `bmad-project-switch` in `workflows/0-context/` for listing, creating, and switching project contexts from the IDE.
- **Inline overrides:** Support `#project:NAME` or `#p:NAME` in invocations for temporary targeting without changing global context.
- **Installer:** Updated to ensure the `0-context/` workflow directory is installed.
- **Compatibility:** 100% backward compatible; default behavior unchanged when no context file is present.
- Implemented with as minimal footprint as I could think of.
### Directory structure (when active)
```
monorepo-root/ 
├── _bmad/ │ 
└── .current_project # Context marker (e.g., "app-alpha") 
├── _bmad-output/ 
│ ├── app-alpha/ 
│ │ ├── planning-artifacts/ 
│ │ └── implementation-artifacts/ 
│ └── app-beta/ 
│ └── planning-artifacts/ 
├── apps/ │ 
├── app-alpha/ 
│ └── app-beta/ 
└── package.json
```
## Testing
- Added `Test Suite 32: Monorepo Config Generation` and `Test Suite 33: Context Skill Discovery` in `test-installation-components.js` to verify monorepo context block extraction and skill discovery.
- All existing tests pass; `npm ci && npm run quality` succeeds.

### CODE REVIEW

File: docs/how-to/monorepo-setup.md [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+A `.current_project` file in your `_bmad/` directory stores the active project name. When set, all artifact output paths are redirected to project-specific subdirectories.
+
+## Quick Start
+
+1. List existing projects by invoking `skill:bmad-project-list`
+2. Create a new project by invoking `skill:bmad-project-new` with a project name
+3. Switch to an existing project by invoking `skill:bmad-project-switch` with a project name
+4. Clear project context by invoking `skill:bmad-project-switch` and entering `CLEAR`
+
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply docs/how-to/monorepo-setup.md]: Fixed in docs/how-to/monorepo-setup.md. The .current_project paragraph now says it is mutable local state, should never be committed, should be ignored in VCS, and can otherwise flip context for other contributors and CI.

File: src/bmm/module.yaml [hunk 2]
```diff
@@ -48,3 +48,35 @@ directories:
...
+#    - REJECT if fails whitelist regex: ^[a-zA-Z0-9._\-/]+$
+#    - On rejection: output "Security Error: Invalid project context" and HALT
+# 4. OVERRIDE PATHS (if valid project_suffix):
+#    - output_folder = {project-root}/_bmad-output/{project_suffix}
+#    - planning_artifacts = {output_folder}/planning-artifacts
+#    - implementation_artifacts = {output_folder}/implementation-artifacts
+#    - project_knowledge = {output_folder}/knowledge
+#    - sprint_status_file = {output_folder}/sprint-status.yaml
+#    - project_name = {project_suffix}
+# 5. If no project_suffix is set, use the default paths from config as-is.
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/module.yaml:72]: valid. I fixed this by standardizing project names as flat single-directory identifiers, not nested paths. Updated in src/bmm/module.yaml, src/bmm/workflows/0-context/bmad-project-new/workflow.md, and src/bmm/workflows/0-context/bmad-project-switch/workflow.md.

File: src/bmm/module.yaml [hunk 3]
```diff
@@ -48,3 +48,35 @@ directories:
...
+monorepo_context:
+  enabled: true
+  context_file: ".current_project"
+  inline_override_patterns:
+    - "#project:"
+    - "#p:"
+  validation:
+    whitelist_regex: "^[a-zA-Z0-9._\\-/]+$"
+    reject_traversal: true
+    reject_absolute: true
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/module.yaml]: valid in principle, but the concrete fix I applied is the same flat-name restriction above. Since / is no longer allowed in project names, forms like foo/, a/./b, and similar path-like inputs are rejected outright. I also tightened the module whitelist in src/bmm/module.yaml

File: src/bmm/workflows/0-context/bmad-project-list/workflow.md [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+## WORKFLOW ARCHITECTURE
+
+### 1. Identify Projects
+
+- Use your file listing or system capabilities to examine the `{project-root}/_bmad-output/` directory.
+- Identify all subdirectories within this path. Each subdirectory represents an available project context, except:
+  - Ignore hidden directories starting with `.`
+  - Ignore standard BMAD artifact directories: `planning-artifacts`, `implementation-artifacts`, `test-artifacts`, `knowledge`, `docs`, `assets`
+- The root `_bmad-output/` directory itself represents the `default (root)` unset project context.
+
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-list/workflow.md]: valid. Fixed in src/bmm/workflows/0-context/bmad-project-list/workflow.md so missing/unreadable _bmad-output/ yields only default (root) instead of a failure.

File: src/bmm/workflows/0-context/bmad-project-new/workflow.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+### 2. Project Creation
+
+1. **Analyze Request**: Determine the requested new project name from the user's initial invocation, such as "new project my-app".
+2. **Wait for Input (If Missing)**: If the user did not provide a project name:
+   - Ask: `What should the new project be called?`
+   - Wait for input.
+3. **Validate Project Name**
+   - **Cleanup**: Remove leading and trailing slashes and any occurrences of `_bmad-output/`.
+   - **Validate - No Traversal**: Reject if path contains `..`.
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-new/workflow.md]: valid. Fixed in src/bmm/workflows/0-context/bmad-project-new/workflow.md to strip only a single leading prefix.

File: src/bmm/workflows/0-context/bmad-project-new/workflow.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+   - **Validate - No Absolute**: Reject if path starts with `/` or a drive letter such as `C:`.
+   - **Validate - Empty/Whitespace**: Reject if empty or only whitespace.
+   - **Validate - Whitelist**: Match against regex `^[-a-zA-Z0-9._/]+$`.
+   - If invalid:
+     - Output: `Error: Invalid project name — must be a relative path and contain only alphanumeric characters, dots, dashes, underscores, or slashes. Traversal (..) is strictly forbidden.`
+     - Halt.
+4. **Check for Existing Project**
+   - Check if `{project-root}/_bmad-output/<sanitized_path>` already exists.
+   - If it already exists:
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-new/workflow.md]: done

File: src/bmm/workflows/0-context/bmad-project-switch/workflow.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+The context file is the default hidden project selector file in `{project-root}/_bmad/` (or as configured in `monorepo_context.context_file`).
+
+### 2. Context Management
+
+1. **Analyze Request**: Determine the requested project name from the user's initial invocation, such as "switch project my-app".
+2. **Wait for Input (If Missing)**: If the user did not provide a project name:
+   - Use your file listing capabilities to examine the `{project-root}/_bmad-output/` directory.
+   - Output a formatted list of the existing projects you found, explicitly noting the `default (root)` project context.
+     Exclude hidden directories and standard BMAD artifact folders: `planning-artifacts`, `implementation-artifacts`, `test-artifacts`, `knowledge`, `docs`, `assets`.
+   - Present options in a numbered list format:
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-switch/workflow.md:25]: done

File: src/bmm/workflows/0-context/bmad-project-switch/workflow.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+   - **Case: Numeric Selection**
+     - Map the number to the corresponding project from your earlier listing.
+     - If number is `0`, treat it as the `CLEAR` case.
+     - If the number is out of range, show an error and ask again.
+     - If valid, use that project name as the path.
+   - **Case: Path Provided** (text input)
+     - **Cleanup**: Remove leading and trailing slashes and any occurrences of `_bmad-output/`.
+     - **Validate - No Traversal**: Reject if path contains `..`.
+     - **Validate - No Absolute**: Reject if path starts with `/` or a drive letter such as `C:`.
+     - **Validate - Empty/Whitespace**: Reject if empty or only whitespace.
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-switch/workflow.md]: Fixed in src/bmm/workflows/0-context/bmad-project-switch/workflow.md so absolute-path validation runs on raw input before cleanup.

File: src/bmm/workflows/0-context/bmad-project-switch/workflow.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+     - **Validate - Whitelist**: Match against regex `^[-a-zA-Z0-9._/]+$`.
+     - **Check Results**
+       - If invalid:
+         - Output: `Error: Invalid project name — must be a relative path and contain only alphanumeric characters, dots, dashes, underscores, or slashes. Traversal (..) is strictly forbidden.`
+         - Halt.
+     - **Validate Existence**: Check if `{project-root}/_bmad-output/<sanitized_path>` exists on disk.
+       - If it does not exist:
+         - Output: `Error: Project <sanitized_path> does not exist. Use PN (Project New) to create it first, or PL (Project List) to see available projects.`
+         - Halt.
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-switch/workflow.md]: Fixed in src/bmm/workflows/0-context/bmad-project-switch/workflow.md.

File: src/bmm/workflows/0-context/bmad-project-switch/workflow.md [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+     - Output: `Project context cleared. Artifacts will go to root _bmad-output/.`
+     - Halt.
+   - **Case: Numeric Selection**
+     - Map the number to the corresponding project from your earlier listing.
+     - If number is `0`, treat it as the `CLEAR` case.
+     - If the number is out of range, show an error and ask again.
+     - If valid, use that project name as the path.
+   - **Case: Path Provided** (text input)
+     - **Cleanup**: Remove leading and trailing slashes and any occurrences of `_bmad-output/`.
+     - **Validate - No Traversal**: Reject if path contains `..`.
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply src/bmm/workflows/0-context/bmad-project-switch/workflow.md]: Fixed in src/bmm/workflows/0-context/bmad-project-switch/workflow.md to strip only a leading prefix.

File: test/test-installation-components.js [hunk 11]
```diff
@@ -1868,6 +1869,118 @@ async function runTests() {
...
+    const generatedConfig32 = await fs.readFile(path.join(tempFixture32, 'bmm', 'config.yaml'), 'utf8');
+    assert(
+      generatedConfig32.includes('# --- Monorepo / Multi-Project Context Support ---'),
+      'Generated bmm config preserves monorepo instruction header',
+    );
+    assert(generatedConfig32.includes('monorepo_context:'), 'Generated bmm config includes monorepo_context YAML');
+    assert(
+      generatedConfig32.includes('inline_override_patterns:') && generatedConfig32.includes('- "#p:"'),
+      'Generated bmm config preserves monorepo override patterns',
+    );
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply test/test-installation-components.js:1917]: Fixed in test/test-installation-components.js by asserting context_file, reject_traversal: true, reject_absolute: true, and the exact whitelist regex are preserved.

File: test/test-installation-components.js [hunk 12]
```diff
@@ -1868,6 +1869,118 @@ async function runTests() {
...
+    assert(
+      !generator33.workflows.some((workflow) => workflow.path.includes('0-context/bmad-project-list')),
+      'List projects skill does not appear in workflows[]',
+    );
+  } catch (error) {
+    assert(false, 'Context skill discovery test succeeds', error.message);
+  } finally {
+    if (tempFixture33) await fs.remove(tempFixture33).catch(() => {});
+  }
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply test/test-installation-components.js:2058]: partially valid. I added useful missing coverage in test/test-installation-components.js, including no-mutable-state assertions and installer helper edge cases. I did not add ManifestGenerator precedence tests because ManifestGenerator has no monorepo precedence logic, and there is no executable JS resolver for inline override vs file vs defaults today. That precedence currently exists as config/workflow behavior, not a shared runtime function.

File: tools/cli/installers/lib/core/installer.js [hunk 13]
```diff
@@ -2067,6 +2067,12 @@ class Installer {
           }
         }
 
+        const staticConfigBlocks = await this.getStaticConfigBlocks(moduleName);
+        if (staticConfigBlocks.length > 0) {
+          yamlContent = yamlContent.trimEnd();
+          yamlContent += `\n\n${staticConfigBlocks.join('\n\n')}\n`;
+        }
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply tools/cli/installers/lib/core/installer.js:2111]: done

File: tools/cli/installers/lib/core/installer.js [hunk 14]
```diff
@@ -2077,6 +2083,65 @@ class Installer {
...
+   */
+  async getStaticConfigBlocks(moduleName) {
+    const moduleSchemaPath = await this.resolveModuleSchemaPath(moduleName);
+    if (!moduleSchemaPath || !(await fs.pathExists(moduleSchemaPath))) {
+      return [];
+    }
+
+    const content = await fs.readFile(moduleSchemaPath, 'utf8');
+    const monorepoBlock = this.extractMonorepoContextBlock(content);
+    return monorepoBlock ? [monorepoBlock] : [];
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply tools/cli/installers/lib/core/installer.js:2137]: Done

File: tools/cli/installers/lib/core/installer.js [hunk 15]
```diff
@@ -2077,6 +2083,65 @@ class Installer {
...
+
+  /**
+   * Resolve the source module.yaml path for a module.
+   * @param {string} moduleName - Module name
+   * @returns {Promise<string|null>} Path to module.yaml or null if not found
+   */
+  async resolveModuleSchemaPath(moduleName) {
+    if (this.configCollector.customModulePaths?.has(moduleName)) {
+      return path.join(this.configCollector.customModulePaths.get(moduleName), 'module.yaml');
+    }
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply tools/cli/installers/lib/core/installer.js:2152]: done

File: tools/cli/installers/lib/core/installer.js [hunk 16]
```diff
@@ -2077,6 +2083,65 @@ class Installer {
...
+  extractMonorepoContextBlock(content) {
+    // Capture from the header comment through to the end of the monorepo_context YAML block.
+    // The block is expected at the end of the file, after any other config sections.
+    const headerIndex = content.indexOf('# --- Monorepo / Multi-Project Context Support ---');
+    if (headerIndex === -1) {
+      return null;
+    }
+
+    const block = content.slice(headerIndex).trimEnd();
+
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply tools/cli/installers/lib/core/installer.js]: done

File: tools/cli/installers/lib/core/installer.js [hunk 17]
```diff
@@ -2077,6 +2083,65 @@ class Installer {
...
+    // Sanity check: the block must contain the structured YAML key
+    if (!block.includes('monorepo_context:')) {
+      return null;
+    }
```
[the remark this answers is not in the document]
Comment (serrnovik) [reply tools/cli/installers/lib/core/installer.js]: done

**Comment by alexeyv** [comment]
Thanks, but. For BMAD proper it's too much extra complexity for what it's worth. What we need to do is figure out a good enough customization story to enable things like that in the customization layer. Closing.
